File: broker.go/internal/bridge/mqttclient/connector.go
package mqttclient

import (
	"context"
	"crypto/tls"
	"fmt"
	"log/slog"
	"strings"
	"sync"
	"time"

	paho "github.com/eclipse/paho.mqtt.golang"
)

// Address is one inbound or outbound mapping rule. Field shapes mirror the
// dashboard's MqttClientAddressInput.
type Address struct {
	Mode        string `json:"mode"`        // SUBSCRIBE | PUBLISH
	RemoteTopic string `json:"remoteTopic"` // topic on the remote broker
	LocalTopic  string `json:"localTopic"`  // topic on the local broker
	QoS         int    `json:"qos"`
	Retain      bool   `json:"retain"`
	// RemovePath: when true, strip the literal prefix of the source-side topic
	// pattern (everything before the first + or #) from the matched topic
	// before mapping to the destination side.
	RemovePath bool `json:"removePath"`
}

// Config is the persisted JSON config (DeviceConfig.Config) for one bridge.
type Config struct {
	BrokerURL         string    `json:"brokerUrl"`
	ClientID          string    `json:"clientId"`
	Username          string    `json:"username,omitempty"`
	Password          string    `json:"password,omitempty"`
	CleanSession      bool      `json:"cleanSession"`
	KeepAlive         int       `json:"keepAlive,omitempty"`
	ConnectionTimeout int       `json:"connectionTimeout,omitempty"`
	ReconnectDelay    int       `json:"reconnectDelay,omitempty"`
	Addresses         []Address `json:"addresses"`

	// Outbound publish buffer for when the remote broker is offline.
	// When BufferEnabled is false (default), publishes that arrive while
	// disconnected are dropped immediately.
	BufferEnabled        bool `json:"bufferEnabled,omitempty"`
	BufferSize           int  `json:"bufferSize,omitempty"`           // 0 → unbounded; recommended ≤ 100_000
	PersistBuffer        bool `json:"persistBuffer,omitempty"`        // true → SQLite-backed; false → in-memory
	DeleteOldestMessages bool `json:"deleteOldestMessages,omitempty"` // true → drop oldest on overflow (FIFO); false → reject new
}

// LocalPublisher is implemented by mochi-mqtt's *Server (Publish).
type LocalPublisher func(topic string, payload []byte, retain bool, qos byte) error

// LocalSubscriber lets the bridge listen to the local broker for outbound
// publishes. Implemented via the in-process pubsub bus.
type LocalSubscriber interface {
	Subscribe(filters []string, buffer int) (id int, ch <-chan LocalMessage)
	Unsubscribe(id int)
}

// LocalMessage is what the bus delivers to the bridge.
type LocalMessage struct {
	Topic   string
	Payload []byte
	QoS     byte
	Retain  bool
}

// Connector is one bridge to one remote broker.
type Connector struct {
	name      string
	cfg       Config
	publisher LocalPublisher
	subBus    LocalSubscriber
	logger    *slog.Logger

	mu     sync.Mutex
	client paho.Client
	subID  int
	stopCh chan struct{}
	buffer Buffer

	IncIn  func()
	IncOut func()
}

func NewConnector(name string, cfg Config, publisher LocalPublisher, subBus LocalSubscriber, logger *slog.Logger) *Connector {
	c := &Connector{
		name: name, cfg: cfg, publisher: publisher, subBus: subBus, logger: logger,
		stopCh: make(chan struct{}),
	}
	if cfg.BufferEnabled {
		c.buffer = c.newBuffer()
	}
	return c
}

// newBuffer constructs the configured outbound buffer; falls back to the
// in-memory variant if the SQLite path can't be opened.
func (c *Connector) newBuffer() Buffer {
	if c.cfg.PersistBuffer {
		path := fmt.Sprintf("./data/bridge_%s.db", c.name)
		b, err := newSqliteBuffer(path, c.cfg.BufferSize, c.cfg.DeleteOldestMessages)
		if err != nil {
			c.logger.Warn("bridge: persistent buffer unavailable, falling back to memory",
				"name", c.name, "path", path, "err", err)
			return newMemoryBuffer(c.cfg.BufferSize, c.cfg.DeleteOldestMessages)
		}
		return b
	}
	return newMemoryBuffer(c.cfg.BufferSize, c.cfg.DeleteOldestMessages)
}

func (c *Connector) Name() string { return c.name }

// Start dials the remote broker and registers inbound/outbound forwarders.
func (c *Connector) Start(ctx context.Context) error {
	opts := paho.NewClientOptions().AddBroker(c.cfg.BrokerURL)
	opts.SetClientID(c.cfg.ClientID)
	if c.cfg.Username != "" {
		opts.SetUsername(c.cfg.Username)
	}
	if c.cfg.Password != "" {
		opts.SetPassword(c.cfg.Password)
	}
	opts.SetCleanSession(c.cfg.CleanSession)
	if c.cfg.KeepAlive > 0 {
		opts.SetKeepAlive(time.Duration(c.cfg.KeepAlive) * time.Second)
	}
	if c.cfg.ConnectionTimeout > 0 {
		opts.SetConnectTimeout(time.Duration(c.cfg.ConnectionTimeout) * time.Second)
	}
	if c.cfg.ReconnectDelay > 0 {
		opts.SetMaxReconnectInterval(time.Duration(c.cfg.ReconnectDelay) * time.Second)
	} else {
		opts.SetMaxReconnectInterval(30 * time.Second)
	}
	opts.SetAutoReconnect(true)
	if strings.HasPrefix(c.cfg.BrokerURL, "ssl://") || strings.HasPrefix(c.cfg.BrokerURL, "tls://") || strings.HasPrefix(c.cfg.BrokerURL, "wss://") {
		opts.SetTLSConfig(&tls.Config{MinVersion: tls.VersionTLS12})
	}

	opts.SetOnConnectHandler(func(_ paho.Client) {
		c.logger.Info("bridge connected", "name", c.name, "url", c.cfg.BrokerURL)
		c.subscribeInbound()
		c.drainBuffer()
	})
	opts.SetConnectionLostHandler(func(_ paho.Client, err error) {
		c.logger.Warn("bridge connection lost", "name", c.name, "err", err)
	})

	client := paho.NewClient(opts)
	tok := client.Connect()
	if !tok.WaitTimeout(5 * time.Second) {
		return fmt.Errorf("bridge %s: connect timeout", c.name)
	}
	if err := tok.Error(); err != nil {
		return fmt.Errorf("bridge %s: %w", c.name, err)
	}
	c.mu.Lock()
	c.client = client
	c.mu.Unlock()

	c.startOutbound(ctx)
	return nil
}

func (c *Connector) subscribeInbound() {
	c.mu.Lock()
	client := c.client
	c.mu.Unlock()
	if client == nil {
		return
	}
	for _, a := range c.cfg.Addresses {
		if !strings.EqualFold(a.Mode, "SUBSCRIBE") {
			continue
		}
		addr := a // capture
		tok := client.Subscribe(addr.RemoteTopic, byte(addr.QoS), func(_ paho.Client, m paho.Message) {
			localTopic := mapInboundTopic(addr, m.Topic())
			if c.IncIn != nil {
				c.IncIn()
			}
			if err := c.publisher(localTopic, m.Payload(), addr.Retain, byte(addr.QoS)); err != nil {
				c.logger.Warn("bridge inbound publish failed", "name", c.name, "topic", localTopic, "err", err)
			}
		})
		if !tok.WaitTimeout(5 * time.Second) {
			c.logger.Warn("bridge subscribe timeout", "name", c.name, "topic", addr.RemoteTopic)
		}
	}
}

// mapInboundTopic maps an incoming topic from the remote broker to the local
// topic to publish under, respecting the address's removePath flag and the
// LocalTopic prefix (if it has no wildcards).
func mapInboundTopic(a Address, remoteTopic string) string {
	t := remoteTopic
	if a.RemovePath {
		if prefix := literalPrefix(a.RemoteTopic); prefix != "" {
			t = strings.TrimPrefix(t, prefix)
			t = strings.TrimPrefix(t, "/")
		}
	}
	if a.LocalTopic == "" || strings.ContainsAny(a.LocalTopic, "+#") {
		return t
	}
	if t == "" {
		return strings.TrimRight(a.LocalTopic, "/")
	}
	return strings.TrimRight(a.LocalTopic, "/") + "/" + t
}

func (c *Connector) startOutbound(ctx context.Context) {
	filters := []string{}
	addrByFilter := map[string]Address{}
	for _, a := range c.cfg.Addresses {
		if !strings.EqualFold(a.Mode, "PUBLISH") {
			continue
		}
		filters = append(filters, a.LocalTopic)
		addrByFilter[a.LocalTopic] = a
	}
	if len(filters) == 0 {
		return
	}
	id, ch := c.subBus.Subscribe(filters, 256)
	c.mu.Lock()
	c.subID = id
	c.mu.Unlock()

	go func() {
		for {
			select {
			case <-ctx.Done():
				return
			case <-c.stopCh:
				return
			case msg, ok := <-ch:
				if !ok {
					return
				}
				addr := pickAddress(addrByFilter, msg.Topic)
				remote := mapOutboundTopic(addr, msg.Topic)
				item := BufferItem{
					Topic: remote, QoS: byte(addr.QoS), Retain: addr.Retain, Payload: msg.Payload,
				}
				c.mu.Lock()
				client := c.client
				buf := c.buffer
				c.mu.Unlock()
				// IsConnectionOpen is the strict variant of IsConnected: it returns
				// false the instant the network drops, whereas IsConnected can
				// briefly remain true while paho's reconnect logic spins up. Without
				// this check, publishes during the gap are silently dropped at QoS 0.
				if client == nil || !client.IsConnectionOpen() {
					if buf != nil {
						if err := buf.Push(item); err != nil {
							c.logger.Warn("bridge buffer push failed", "name", c.name, "err", err)
						}
					}
					continue
				}
				if err := c.publishOne(client, item); err != nil {
					// Failed to publish while supposedly connected — buffer
					// it so the next OnConnect can drain it.
					if buf != nil {
						if perr := buf.Push(item); perr != nil {
							c.logger.Warn("bridge buffer push failed after publish error",
								"name", c.name, "err", perr)
						}
					}
				}
			}
		}
	}()
}

// publishOne sends a single buffered or fresh item to the remote broker via
// paho. Returns the publish-token error so the caller can decide to re-buffer
// on failure.
func (c *Connector) publishOne(client paho.Client, item BufferItem) error {
	if c.IncOut != nil {
		c.IncOut()
	}
	tok := client.Publish(item.Topic, item.QoS, item.Retain, item.Payload)
	if !tok.WaitTimeout(5 * time.Second) {
		return fmt.Errorf("publish timeout topic=%s", item.Topic)
	}
	return tok.Error()
}

// drainBuffer is called from the OnConnect handler. It walks the outbound
// buffer FIFO and re-publishes each entry; on the first publish failure it
// stops and leaves the rest in the buffer for the next reconnect.
func (c *Connector) drainBuffer() {
	c.mu.Lock()
	buf := c.buffer
	client := c.client
	c.mu.Unlock()
	if buf == nil || client == nil {
		return
	}
	if buf.Len() == 0 {
		return
	}
	c.logger.Info("bridge: draining outbound buffer", "name", c.name, "pending", buf.Len())
	if err := buf.Drain(func(item BufferItem) error { return c.publishOne(client, item) }); err != nil {
		c.logger.Warn("bridge: buffer drain interrupted", "name", c.name, "err", err)
	}
}

func pickAddress(filters map[string]Address, topic string) Address {
	for filter, a := range filters {
		if filter == "#" || matchTopic(filter, topic) {
			return a
		}
	}
	return Address{}
}

// mapOutboundTopic maps a locally-published topic to the topic to forward to
// the remote broker.
func mapOutboundTopic(a Address, localTopic string) string {
	t := localTopic
	if a.RemovePath {
		if prefix := literalPrefix(a.LocalTopic); prefix != "" {
			t = strings.TrimPrefix(t, prefix)
			t = strings.TrimPrefix(t, "/")
		}
	}
	if a.RemoteTopic == "" || strings.ContainsAny(a.RemoteTopic, "+#") {
		return t
	}
	if t == "" {
		return strings.TrimRight(a.RemoteTopic, "/")
	}
	return strings.TrimRight(a.RemoteTopic, "/") + "/" + t
}

// literalPrefix returns the longest literal prefix of an MQTT topic pattern —
// i.e. everything up to but not including the first wildcard segment.
//
//	"sensor/#"        → "sensor"
//	"a/b/+/c"         → "a/b"
//	"+/x"             → ""
//	"plain/topic"     → "plain/topic"
func literalPrefix(pattern string) string {
	parts := strings.Split(pattern, "/")
	for i, p := range parts {
		if p == "+" || p == "#" {
			return strings.Join(parts[:i], "/")
		}
	}
	return pattern
}

func (c *Connector) Stop() {
	c.mu.Lock()
	defer c.mu.Unlock()
	select {
	case <-c.stopCh:
	default:
		close(c.stopCh)
	}
	if c.subID != 0 && c.subBus != nil {
		c.subBus.Unsubscribe(c.subID)
		c.subID = 0
	}
	if c.client != nil {
		c.client.Disconnect(250)
		c.client = nil
	}
	if c.buffer != nil {
		_ = c.buffer.Close()
		c.buffer = nil
	}
}

// BufferLen exposes the current outbound buffer depth (0 if buffering is
// disabled). Used by tests and metrics.
func (c *Connector) BufferLen() int {
	c.mu.Lock()
	buf := c.buffer
	c.mu.Unlock()
	if buf == nil {
		return 0
	}
	return buf.Len()
}

func matchTopic(pattern, topic string) bool {
	pp := strings.Split(pattern, "/")
	tt := strings.Split(topic, "/")
	for i, p := range pp {
		if p == "#" {
			return true
		}
		if i >= len(tt) {
			return false
		}
		if p == "+" {
			continue
		}
		if p != tt[i] {
			return false
		}
	}
	return len(pp) == len(tt)
}

File: broker.rs/src/bridge/mqttclient/connector.rs
use crate::pubsub::Bus as PubsubBus;
use crate::stores::topic::match_topic;
use parking_lot::Mutex;
use rumqttc::{AsyncClient, Event, EventLoop, MqttOptions, Packet, QoS, Transport};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use std::time::Duration;
use tokio::task::JoinHandle;

/// One inbound or outbound mapping rule. Mirrors the dashboard's
/// MqttClientAddressInput.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Address {
    #[serde(default)]
    pub mode: String, // SUBSCRIBE | PUBLISH
    #[serde(default, rename = "remoteTopic")]
    pub remote_topic: String,
    #[serde(default, rename = "localTopic")]
    pub local_topic: String,
    #[serde(default)]
    pub qos: i32,
    #[serde(default)]
    pub retain: bool,
    /// removePath: when true, strip the literal prefix of the source-side
    /// pattern (everything before the first `+` or `#`) from the matched
    /// topic before mapping to the destination side.
    #[serde(default, rename = "removePath")]
    pub remove_path: bool,
}

/// Persisted JSON config (DeviceConfig.config) for one bridge.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Config {
    #[serde(default, rename = "brokerUrl")]
    pub broker_url: String,
    #[serde(default, rename = "clientId")]
    pub client_id: String,
    #[serde(default)]
    pub username: String,
    #[serde(default)]
    pub password: String,
    #[serde(default, rename = "cleanSession")]
    pub clean_session: bool,
    #[serde(default, rename = "keepAlive")]
    pub keep_alive: i32,
    #[serde(default, rename = "connectionTimeout")]
    pub connection_timeout: i32,
    #[serde(default, rename = "reconnectDelay")]
    pub reconnect_delay: i32,
    #[serde(default)]
    pub addresses: Vec<Address>,
}

/// Local publish callback — invoked when a remote message comes in.
pub type LocalPublishFn = Arc<dyn Fn(&str, Vec<u8>, bool, u8) + Send + Sync>;

/// Local message delivered by the in-process pubsub bus to the bridge.
#[derive(Debug, Clone)]
pub struct LocalMessage {
    pub topic: String,
    pub payload: Vec<u8>,
    pub qos: u8,
    pub retain: bool,
}

/// Connector — one bridge to one remote broker.
pub struct Connector {
    pub name: String,
    cfg: Config,
    publisher: LocalPublishFn,
    pubsub: Arc<PubsubBus>,
    handles: Mutex<Vec<JoinHandle<()>>>,
    client: Mutex<Option<AsyncClient>>,
}

impl Connector {
    pub fn new(name: String, cfg: Config, publisher: LocalPublishFn, pubsub: Arc<PubsubBus>) -> Arc<Self> {
        Arc::new(Self {
            name, cfg, publisher, pubsub,
            handles: Mutex::new(vec![]),
            client: Mutex::new(None),
        })
    }

    pub async fn start(self: Arc<Self>) -> anyhow::Result<()> {
        let url = self.cfg.broker_url.clone();
        let (host, port, transport) = parse_broker_url(&url)?;

        let client_id = if self.cfg.client_id.is_empty() {
            format!("monstermq-{}", &self.name)
        } else { self.cfg.client_id.clone() };

        let mut opts = MqttOptions::new(client_id, host, port);
        opts.set_keep_alive(Duration::from_secs(self.cfg.keep_alive.max(30) as u64));
        opts.set_clean_session(self.cfg.clean_session);
        opts.set_transport(transport);
        if !self.cfg.username.is_empty() {
            opts.set_credentials(&self.cfg.username, &self.cfg.password);
        }

        let (client, mut eventloop): (AsyncClient, EventLoop) = AsyncClient::new(opts, 256);
        *self.client.lock() = Some(client.clone());

        // Inbound: subscribe to remote topics for SUBSCRIBE addresses.
        for a in self.cfg.addresses.iter().filter(|a| a.mode.eq_ignore_ascii_case("SUBSCRIBE")) {
            let qos = match a.qos { 0 => QoS::AtMostOnce, 1 => QoS::AtLeastOnce, _ => QoS::ExactlyOnce };
            let _ = client.subscribe(&a.remote_topic, qos).await;
        }

        let me_in = self.clone();
        let h = tokio::spawn(async move {
            loop {
                match eventloop.poll().await {
                    Ok(Event::Incoming(Packet::Publish(p))) => {
                        if let Some(addr) = me_in.find_subscribe_addr(&p.topic) {
                            let local_topic = map_inbound_topic(&addr, &p.topic);
                            (me_in.publisher)(&local_topic, p.payload.to_vec(), addr.retain, p.qos as u8);
                        }
                    }
                    Ok(_) => {}
                    Err(e) => {
                        tracing::warn!(name = me_in.name, err = ?e, "bridge poll error");
                        tokio::time::sleep(Duration::from_secs(2)).await;
                    }
                }
            }
        });
        self.handles.lock().push(h);

        // Outbound: subscribe to the local pubsub bus for PUBLISH addresses.
        let pub_filters: Vec<String> = self.cfg.addresses.iter()
            .filter(|a| a.mode.eq_ignore_ascii_case("PUBLISH"))
            .map(|a| a.local_topic.clone())
            .collect();
        if !pub_filters.is_empty() {
            let mut sub = self.pubsub.subscribe(pub_filters, 256);
            let me_out = self.clone();
            let client_out = client.clone();
            let h = tokio::spawn(async move {
                while let Some(msg) = sub.rx.recv().await {
                    if let Some(addr) = me_out.find_publish_addr(&msg.topic_name) {
                        let remote = map_outbound_topic(&addr, &msg.topic_name);
                        let qos = match addr.qos { 0 => QoS::AtMostOnce, 1 => QoS::AtLeastOnce, _ => QoS::ExactlyOnce };
                        let _ = client_out.publish(&remote, qos, addr.retain, msg.payload.clone()).await;
                    }
                }
            });
            self.handles.lock().push(h);
        }
        Ok(())
    }

    pub async fn stop(&self) {
        // Drop the guard before awaiting (parking_lot guards are !Send).
        let client = self.client.lock().take();
        if let Some(c) = client {
            let _ = c.disconnect().await;
        }
        let handles: Vec<_> = self.handles.lock().drain(..).collect();
        for h in handles { h.abort(); }
    }

    fn find_subscribe_addr(&self, remote_topic: &str) -> Option<Address> {
        self.cfg.addresses.iter()
            .filter(|a| a.mode.eq_ignore_ascii_case("SUBSCRIBE"))
            .find(|a| match_topic(&a.remote_topic, remote_topic))
            .cloned()
    }

    fn find_publish_addr(&self, local_topic: &str) -> Option<Address> {
        self.cfg.addresses.iter()
            .filter(|a| a.mode.eq_ignore_ascii_case("PUBLISH"))
            .find(|a| match_topic(&a.local_topic, local_topic))
            .cloned()
    }
}

fn parse_broker_url(url: &str) -> anyhow::Result<(String, u16, Transport)> {
    let (scheme, rest) = url.split_once("://").ok_or_else(|| anyhow::anyhow!("invalid brokerUrl: {}", url))?;
    let (host, port_s) = match rest.split_once(':') {
        Some((h, p)) => (h.to_string(), p.parse::<u16>().unwrap_or(0)),
        None => (rest.to_string(), 0),
    };
    let (default_port, transport) = match scheme {
        "tcp" | "mqtt" => (1883u16, Transport::tcp()),
        "ssl" | "tls" | "mqtts" => (8883, Transport::tls_with_default_config()),
        // TODO(broker.rs): WS / WSS bridge transports — requires the `websocket`
        // feature on rumqttc and a different Transport constructor.
        "ws" | "wss" => anyhow::bail!("ws/wss bridge transport not yet implemented in broker.rs"),
        _ => anyhow::bail!("unsupported brokerUrl scheme: {}", scheme),
    };
    let port = if port_s == 0 { default_port } else { port_s };
    Ok((host, port, transport))
}

/// Maps an incoming topic from the remote broker to the local topic to publish
/// under, respecting the address's removePath flag and the LocalTopic prefix
/// (if it has no wildcards).
pub fn map_inbound_topic(a: &Address, remote_topic: &str) -> String {
    let mut t = remote_topic.to_string();
    if a.remove_path {
        let prefix = literal_prefix(&a.remote_topic);
        if !prefix.is_empty() {
            t = t.strip_prefix(&prefix).unwrap_or(&t).to_string();
            t = t.strip_prefix('/').unwrap_or(&t).to_string();
        }
    }
    if a.local_topic.is_empty() || a.local_topic.contains('+') || a.local_topic.contains('#') {
        return t;
    }
    if t.is_empty() {
        return a.local_topic.trim_end_matches('/').to_string();
    }
    format!("{}/{}", a.local_topic.trim_end_matches('/'), t)
}

pub fn map_outbound_topic(a: &Address, local_topic: &str) -> String {
    let mut t = local_topic.to_string();
    if a.remove_path {
        let prefix = literal_prefix(&a.local_topic);
        if !prefix.is_empty() {
            t = t.strip_prefix(&prefix).unwrap_or(&t).to_string();
            t = t.strip_prefix('/').unwrap_or(&t).to_string();
        }
    }
    if a.remote_topic.is_empty() || a.remote_topic.contains('+') || a.remote_topic.contains('#') {
        return t;
    }
    if t.is_empty() {
        return a.remote_topic.trim_end_matches('/').to_string();
    }
    format!("{}/{}", a.remote_topic.trim_end_matches('/'), t)
}

/// Longest literal prefix of an MQTT topic pattern — i.e. everything up to but
/// not including the first wildcard segment.
///
/// `"sensor/#"        → "sensor"`
/// `"a/b/+/c"         → "a/b"`
/// `"+/x"             → ""`
/// `"plain/topic"     → "plain/topic"`
pub fn literal_prefix(pattern: &str) -> String {
    let parts: Vec<&str> = pattern.split('/').collect();
    for (i, p) in parts.iter().enumerate() {
        if *p == "+" || *p == "#" {
            return parts[..i].join("/");
        }
    }
    pattern.to_string()
}
